plutus-ledger/src/Ledger/Tx.hs

```diff
@@ -75,99 +73,119 @@ import Data.Proxy (Proxy (Proxy))
 import Data.Set (Set)
 import Data.Set qualified as Set
 import GHC.Generics (Generic)
-import Ledger.Address (Address, PaymentPubKey, StakePubKey, pubKeyAddress)
+import Ledger.Address (PaymentPubKey, StakePubKey, pubKeyAddress)
 import Ledger.Crypto (Passphrase, signTx, signTx', toPublicKey)
 import Ledger.Orphans ()
 import Ledger.Slot (SlotRange)
 import Ledger.Tx.CardanoAPI (SomeCardanoApiTx (SomeTx), ToCardanoError (..))
 import Ledger.Tx.CardanoAPI qualified as CardanoAPI
 import Ledger.Tx.Internal hiding (updateUtxoCollateral)
-import Plutus.Script.Utils.V1.Scripts (datumHash)
+import Plutus.Script.Utils.V1.Scripts (datumHash, validatorHash)
 import Plutus.V1.Ledger.Api qualified as V1
-import Plutus.V2.Ledger.Api qualified as V2
 -- for re-export
+import Plutus.V1.Ledger.Address qualified as V1
 import Plutus.V1.Ledger.Tx
 import Plutus.V1.Ledger.Tx qualified as V1.Tx
 import Prettyprinter (Pretty (pretty), braces, colon, hang, nest, viaShow, vsep, (<+>))
 
 type PrivateKey = Crypto.XPrv
 
--- | Transaction output that comes from a chain index query.
---
--- It is defined here instead of the plutus-chain-index because plutus-ledger
--- uses that datatype, and plutus-ledger can't depend on plutus-chain-index
--- because of a cyclic dependency.
---
--- This datatype was created in order to be used in
--- 'Ledger.Constraints.processConstraint', specifically with the constraints
--- 'MustSpendPubKeyOutput' and 'MustSpendScriptOutput'.
-data ChainIndexTxOut =
-    PublicKeyChainIndexTxOut {
-      -- | Address of the transaction output. The address is protected by a
-      -- public key hash.
-      _ciTxOutAddress         :: Address,
+-- | Transaction output represented off-chain
+-- add more
+data OffChainTxOut =
+    PublicKeyOffChainTxOut' {
+      -- | Hash of the public key protecting the transaction output.
+      ocTxOutAddress            :: V1.PubKeyHash,
       -- | Value of the transaction output.
-      _ciTxOutValue           :: V1.Value,
-      -- | Optional datum attached to the transaction output.
-      _ciTxOutPublicKeyDatum  :: V2.OutputDatum,
+      ocTxOutValue              :: V1.Value,
+      -- | Optional datum hash attached to the transaction output.
+      ocTxOutPublicKeyDatumHash :: Maybe V1.DatumHash,
+      -- | Optional datum attached to the transaction output, either found
+      -- inline or resolved from its hash offchain.
+      ocTxOutDatum              :: Maybe V1.Datum,
```

Comment:
Why into 2 separate fields instead of a Tuple?

Reply:
Because you might have a `DatumHash` but not a `Datum`. It would be tighter to use `Maybe (DatumHash, Maybe Datum)` since you cannot have a Datum without its hash.